### PR TITLE
/tg/ Status Effects Part 2 - datum, KD, KO, Stuns

### PR DESCRIPTION
**Pull request by fira**
# About the pull request

Part 2 - this includes porting the actual status_effect datum, modifying it to fit our purposes by backing it with timers similarly to old system, and finally implementing KD, KO and Stun with it.

This contains Part 1 PR (#4828) so if you want to take a look at it I'd advise checking the last commits or setting up a compare between both branches.

# Explain why it's good for the game
Predictable status timers. Current ones are bogus in their handling of "life tick correction" and will "stack" time even when they're not supposed to.

Also provides a more robust backend for general effects, and integrates status effects into it.

# Testing Photographs and Procedure

Summary testing of buckling interactions, explosion knock times, crawling, resting. Will have to be expanded once part 1 is ready

# Changelog
:cl:
add: Added Buckled, Handcuffed and Legcuffed screen alerts
code: Ported /tg/ status effects backend, modified with timers to let effects end at appropriate time
code: Stun, Knockdown and Knockout now use the new effects backend
balance: Due to backend change, all KO/KD/Stuns may behave differently timing wise. This is of course subject to adjustments.
balance: Endurance is now set at 8% effect duration reduction per level above 1. However it now compounds with species bonus. Feel free to adjust.
balance: Knockdowns are not inherently incapacitating anymore and many sources of it have been updated to also stun to make up for it.
fix: KO/KD/Stuns do not artificially and randomly ''stack'' due to incorrect timer offset calculation anymore.
fix: Stuns now correctly apply Stun reduction values instead of Knockdown reductions.
fix: Crawling can now be interrupted by a normal move, if you are fit enough to do so.
/:cl:

### CODE REVIEW

**Comment by fira** [comment]
FWIW most of the sprites are actually unused, they didn't blend in well with the rest of the UI, so i just left cuff/buckle ones. Might need to cleanup

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by fira** [comment]
Regression to address, humans can crawl away from explosions AGAIN

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: code/datums/status_effects/_status_effect.dm [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+		deltimer(timerid)
+		timerid = null
+
+/// Timer invocation callback to end the effect
+/datum/status_effect/proc/timer_callback()
+	if(timerid)
+		timerid = null
+	qdel(src) // shrimple as that
+
+/// Called whenever the effect is applied in on_created
```
[the remark this answers is not in the document]
Comment (fira) [reply code/datums/status_effects/_status_effect.dm:142]: That's actually a /tg/ mistake but yeah I'll change it

**Review by zzzmike — COMMENTED** [review]
image For lines 277 and 287 - I guess all weakens need to be changed to knockdown and/or stun? This should resolve my comment on https://github.com/cmss13-devs/cmss13/issues/5074

**Comment by fira** [comment]
Yup exactly

Right now Stun is preventing you from taking actions + moving, so Stun + KnockDown is equivalent to old KnockDown functionality. New KD just prevents you from standing and moving (important distinction due to crawling)

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/atom/movable/screen/alert/status_effect/stun
+	name = "Stunned"
+	desc = "You are incapacitated. You may not move or act."
+	icon_state = ALERT_INCAPACITATED
+
+
+//KNOCKDOWN
+/datum/status_effect/incapacitating/knockdown
+	id = "knockdown"
+//	alert_type = /atom/movable/screen/alert/status_effect/knockdown
```
Comment (harryob) [inline code/datums/status_effects/debuffs/debuffs.dm:48]: still pending sprites or is that gonna be a later pr

File: code/modules/mob/living/living_health_procs.dm [hunk 3]
```diff
@@ -174,103 +170,97 @@
...
+		S.adjust_duration(amount)
+	else if(amount > 0)
+		S = apply_status_effect(/datum/status_effect/incapacitating/knockdown, amount)
+	return S
+
+/* KnockOut (Unconscious) */
+/// Overridable handler to adjust the numerical value of status effects. Expand as needed
+/mob/living/proc/GetKnockOutDuration(amount)
+	return amount * GLOBAL_STATUS_MULTIPLIER
+/mob/living/proc/IsKnockOut()
```
Comment (harryob) [inline code/modules/mob/living/living_health_procs.dm:223]:
```suggestion

/mob/living/proc/IsKnockOut()
```

File: code/modules/mob/living/living_health_procs.dm [hunk 4]
```diff
@@ -174,103 +170,97 @@
...
+	return has_status_effect(/datum/status_effect/incapacitating/unconscious)
+/mob/living/proc/AmountKnockOut() //How much time remains - scaled by GLOBAL_STATUS_MULTIPLIER (normally in multiples of legacy 2 seconds)
```
Comment (harryob) [inline code/modules/mob/living/living_health_procs.dm:225]:
```suggestion

/// How much time remains - scaled by GLOBAL_STATUS_MULTIPLIER (normally in multiples of legacy 2 seconds)
/mob/living/proc/AmountKnockOut() 
```

File: code/modules/mob/living/living_health_procs.dm [hunk 5]
```diff
@@ -174,103 +170,97 @@
...
+	var/datum/status_effect/incapacitating/unconscious/S = IsKnockOut()
+	if(S)
+		return S.get_duration_left() / GLOBAL_STATUS_MULTIPLIER
+	return 0
```
Comment (harryob) [inline code/modules/mob/living/living_health_procs.dm:229]:
```suggestion
	return 0

```

File: code/modules/mob/living/living_health_procs.dm [hunk 6]
```diff
@@ -174,103 +170,97 @@
...
+	if(amount <= 0)
+		if(S)
+			qdel(S)
+	else
+		if(S)
+			S.update_duration(amount)
+		else
+			S = apply_status_effect(/datum/status_effect/incapacitating/unconscious, amount)
+	return S
+/mob/living/proc/AdjustKnockOut(amount, ignore_canstun = FALSE) //Adds to remaining duration
```
Comment (harryob) [inline code/modules/mob/living/living_health_procs.dm:254]:
```suggestion

/mob/living/proc/AdjustKnockOut(amount, ignore_canstun = FALSE) //Adds to remaining duration
```

File: code/modules/mob/living/living_health_procs.dm [hunk 7]
```diff
@@ -174,103 +170,97 @@
...
+	if(!(status_flags & CANKNOCKOUT))
+		return
+	amount = GetKnockOutDuration(amount)
+	var/datum/status_effect/incapacitating/unconscious/S = IsKnockOut()
+	if(S)
+		S.update_duration(amount, increment = TRUE)
+	else if(amount > 0)
+		S = apply_status_effect(/datum/status_effect/incapacitating/unconscious, amount)
+	return S
+/mob/living/proc/SetKnockOut(amount, ignore_canstun = FALSE) //Sets remaining duration
```
Comment (harryob) [inline code/modules/mob/living/living_health_procs.dm:240]:
```suggestion

/mob/living/proc/SetKnockOut(amount, ignore_canstun = FALSE) //Sets remaining duration
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: code/modules/mob/living/carbon/xenomorph/XenoProcs.dm [hunk 8]
```diff
@@ -317,7 +317,8 @@
 	visible_message(SPAN_DANGER("[src] [pounceAction.ability_name] onto [M]!"), SPAN_XENODANGER("You [pounceAction.ability_name] onto [M]!"), null, 5)
 
 	if (pounceAction.knockdown)
-		M.apply_effect(pounceAction.knockdown_duration, WEAKEN)
+		M.KnockDown(pounceAction.knockdown_duration)
+		M.Stun(pounceAction.knockdown_duration) // To replicate legacy bheavior. Otherwise M39 Armbrace users for example can still shoot
```
Comment (private-tristan) [inline code/modules/mob/living/carbon/xenomorph/XenoProcs.dm:321]: bheavior 😭

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	. = ..()
+	if(!.)
+		return
+	ADD_TRAIT(owner, TRAIT_KNOCKEDOUT, TRAIT_STATUS_EFFECT(id))
+
+/datum/status_effect/incapacitating/unconscious/on_remove()
+	REMOVE_TRAIT(owner, TRAIT_KNOCKEDOUT, TRAIT_STATUS_EFFECT(id))
+	return ..()
+
+// /datum/status_effect/incapacitating/unconscious/tick(seconds_between_ticks)
```
Comment (Zonespace27) [inline code/datums/status_effects/debuffs/debuffs.dm:101]: Commented code

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 10]
```diff
@@ -0,0 +1,108 @@
+//Largely negative status effects go here, even if they have small benificial effects
+//STUN EFFECTS
+/datum/status_effect/incapacitating
+	tick_interval = -1
+	status_type = STATUS_EFFECT_REPLACE
+	alert_type = null
+	remove_on_fullheal = TRUE
+//	heal_flag_necessary = HEAL_CC_STATUS
```
Comment (Zonespace27) [inline code/datums/status_effects/debuffs/debuffs.dm:8]: .

File: code/datums/status_effects/_status_effect.dm [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+	return 0
+
+/// Signal proc for [COMSIG_LIVING_REJUVENATED] to remove us on fullheal
+/datum/status_effect/proc/remove_effect_on_heal(datum/source, heal_flags)
+	SIGNAL_HANDLER
+
+	if(!remove_on_fullheal)
+		return
+
+//	if(!heal_flag_necessary || (heal_flags & heal_flag_necessary))
```
Comment (Zonespace27) [inline code/datums/status_effects/_status_effect.dm:211]: .

File: code/datums/status_effects/stacking_effect.dm [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+		qdel(src)
+	else
+		add_stacks(-stack_decay)
+		stack_decay_effect()
+
+/// Add (or remove) [stacks_added] stacks to our current stack count.
+/datum/status_effect/stacking/proc/add_stacks(stacks_added)
+	if(stacks_added > 0 && !can_gain_stacks())
+		return FALSE
+//o	owner.cut_overlay(status_overlay)
```
Comment (Zonespace27) [inline code/datums/status_effects/stacking_effect.dm:94]: .

**Review by Zonespace27 — APPROVED** [review]

**Comment by fira** [comment]
I cut down overlay integration & some other commented code 

Kept the commented rejuvenation flags since it's something we might want short term when adding more effects

**Pull request merged**